File: frontend/src/types/model-registry.ts
export type RegisterModelSource = "hf" | "local" | "s3";
export type RegisterModelDtype = "bfloat16" | "float16" | "float32" | "int8" | "int4";

export interface ModelRegistryEntry {
  readonly name: string;
  readonly source: string;
  readonly isPinned: boolean;
  readonly params: string | null;
  readonly context: string | null;
  readonly license: string | null;
  readonly path: string | null;
  readonly dtype: string | null;
}

export interface RegisterModelEntryRequest {
  readonly name: string;
  readonly source: RegisterModelSource;
  readonly path: string;
  readonly dtype: RegisterModelDtype;
  readonly isPinned: boolean;
}

File: backend/app/schemas/model_registry.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel

RegisterModelSource = Literal["hf", "local", "s3"]
RegisterModelDtype = Literal["bfloat16", "float16", "float32", "int8", "int4"]


class ModelRegistryEntryResponse(BaseModel):
    name: str
    source: str
    is_pinned: bool
    params: str | None = None
    context: str | None = None
    license: str | None = None
    path: str | None = None
    dtype: str | None = None

    model_config = {"from_attributes": True}


class ModelRegistryResponse(BaseModel):
    entries: list[ModelRegistryEntryResponse]


class RegisterModelEntryRequest(BaseModel):
    name: str
    source: RegisterModelSource
    path: str
    dtype: RegisterModelDtype
    is_pinned: bool
